assemblyline_v4_service/updater/client.py

```diff
@@ -10,7 +12,171 @@
 SIGNATURE_UPDATE_BATCH = int(os.environ.get('SIGNATURE_UPDATE_BATCH', '1000'))
 
 
-class SyncableSignature(SignatureClient):
+class SyncableBadlistClient(BadlistClient):
+    def __init__(self, datastore, config=None):
+        super().__init__(datastore, config)
+        self.sync = False
+
+    def add_update_many(self, data: List[Union[dict, BadlistModel]]) -> Dict[str, Any]:
+        # This version of the API allows to sync badlist items with the system by making direct changes to the datastore
+        # Items that no longer exist at the source will be removed from `sources` to maintain active synchronicity
+        # amongst multiple sources.
+        # If there are no more sources actively blocking the item, then the item will be DISABLED but users can always
+        # re-deploy item if desired
+
+        current_ids = set()
+        source = None
+
+        # Iterate over the list of signatures given
+        for i, d in enumerate(data):
+            if isinstance(d, BadlistModel):
+                d = d.as_primitives()
+
+            if not source:
+                # Set the source name
+                source = data["sources"][0]["name"]
+
+            if self.sync:
+                # Compute the expected badlist ID and add it to the list
+                current_ids.add(self._preprocess_object(d))
+
+            # Update with JSON-friendly version of data to be sent to API
+            data[i] = d
+
+        if self.sync:
+            # Get the list of items that currently existing in the system for the source
+            existing_ids = set(
+                [
+                    i["id"]
+                    for i in self.datastore.badlist.stream_search(
+                        f"sources.name:{source}", fl="id", as_obj=False
+                    )
+                ]
+            )
+
+            # Find the IDs that don't exist at this source anymore and remove the source from the source list
+            for missing_id in existing_ids - current_ids:
+                missing_item: BadlistModel = self.datastore.badlist.get(missing_id)
+                missing_item.sources = [
+                    s
+                    for s in missing_item.sources
```

Comment:
Atho the item will be disabled, this will create an item with no source which I don't think we should do. We could at least keep the last source... I believe some of the code for badlist assume there is at least one source.